### PR TITLE
improve independence of vision components

### PR DESCRIPTION
**Pull request by epwalsh**
This this does a couple things.

First, I added an optional "extra" to the `setup.py` called "vision", where we can specify vision-only requirements like `torchvision` and `detectron2`, except not `detectron2` because that has to be installed separately and prior to installing AllenNLP since it requires that `torch` is installed first 🙃

The other thing the PR does is sort of de-couple all of the vision-specific submodules so that AllenNLP won't break when the vision dependencies are missing, i.e. when someone just installs AllenNLP with `pip install allennlp`, as opposed to `pip install allennlp[vision]`.

This seems a little hacky with `try-except` statements around imports, but I don't know a better way aside from rewriting the entire library in Rust where this could be trivially solved with optional package features.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -20,7 +20,11 @@ WORKDIR /stage/allennlp
 
 # Install the wheel of AllenNLP.
 COPY dist dist/
-RUN pip install $(ls dist/*.whl)
+RUN pip install $(ls dist/*.whl)[vision]
+
+# Install detectron2.
+RUN pip install detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cu102/torch1.7/index.html
```
Comment (dirkgr) [inline Dockerfile:26]: This line has the CUDA version and the torch version built in. Is there a way to not be dependent on that?

File: tests/training/metrics/auc_test.py [hunk 2]
```diff
@@ -100,6 +100,7 @@ def test_distributed_auc(self):
         labels = torch.randint(3, 5, (8,), dtype=torch.long)
         # We make sure that the positive label is always present.
         labels[0] = 4
+        labels[4] = 4
```
Comment (dirkgr) [inline tests/training/metrics/auc_test.py:103]: This is unrelated?

**Review by dirkgr — APPROVED** [review]

**Review by AkshitaB — APPROVED** [review]

File: tests/training/metrics/auc_test.py (hunk 2 above)
Comment (epwalsh) [reply tests/training/metrics/auc_test.py:103]: Unrelated. Might need this on master too actually.

File: Dockerfile (hunk 1 above)
Comment (epwalsh) [reply Dockerfile:26]: Yea, good point. We could (and should) make that a build variable.

Comment (epwalsh) [reply Dockerfile:26]: Once https://github.com/allenai/allennlp/pull/4796 goes through I'll add a build ARG for detectron.
